README.md
# react-tray

[![Build Status](https://travis-ci.org/instructure-react/react-tray.svg?branch=master)](https://travis-ci.org/instructure-react/react-tray)

An accessible tray component useful for navigation menus
See example at [http://instructure-react.github.io/react-tray](http://instructure-react.github.io/react-tray)

## Usage

```js
var React = require('react');
var Tray = require('react-tray');

var App = React.createClass({
  getInitialState: function () {
    return {
      isTrayOpen: false
    };
  },

  openTray: function () {
    this.setState({
      isTrayOpen: true
    });
  },

  closeTray: function () {
    this.setState({
      isTrayOpen: false
    });
  },


  render: function () {
    return (
      <div>
        <ul role="menu" className="navigation">
          <li role="menuitem">
            <a tabIndex={0}
                role="button"
                aria-haspopup={true}
                onClick={this.handleNavClick}>Menu</a>
          </li>
        </ul>
        <Tray isOpen={this.state.isTrayOpen}
              onBlur={this.closeTray}
              closeTimeoutMS={150}
        >
          <h1>Tray Content</h1>
          <div>Learn to drive and everything.</div>
        </Tray>
      </div>
    );
  }
});

React.render(<App/>, document.getElementById('content'));
```



CHANGELOG.md
v2.0.4 - Tue, 25 Oct 2016 20:21:50 GMT
--------------------------------------

- 


v2.0.3 - Tue, 25 Oct 2016 20:11:21 GMT
--------------------------------------

- 


v2.0.2 - Mon, 17 Oct 2016 15:55:11 GMT
--------------------------------------
This commit mostly addresses the `dist/` versions for use in AMD builds.  They are now named properly fixing the issue with https://github.com/instructure-react/react-tray/pull/10
- [c0a6782](../../commit/c0a6782) Updated rackt-cli build/release tooling version


v2.0.1 - Wed, 13 Jul 2016 22:55:21 GMT
--------------------------------------

- [36e3e55](../../commit/36e3e55) [fixed] Split toggleAriaHidden internal method to two methods


v2.0.0 - Wed, 13 Jul 2016 14:42:45 GMT
--------------------------------------

- [f3b4731](../../commit/f3b4731) [changed] Move elementToFocus prop to getElementToFocus (#9)
- [7a316ff](../../commit/7a316ff) [added] Prop for adding aria-hidden to application element (#8)


v1.1.0 - Mon, 11 Jul 2016 22:24:30 GMT
--------------------------------------

- [d860b32](../../commit/d860b32) [fixed] Failing test for elementToFocus
- [3300f2c](../../commit/3300f2c) [added] Add elementToFocus prop (#5)
- [afe8d85](../../commit/afe8d85) [added] onOpen callback


v1.0.1 - Fri, 08 Jul 2016 22:00:48 GMT
--------------------------------------

- [a374410](../../commit/a374410) [fixed] make default maintainFocus prob actually be true


v1.0.0 - Fri, 08 Jul 2016 21:20:29 GMT
--------------------------------------

- [720c1ce](../../commit/720c1ce) [changed] Maintain focus inside of tray


v0.3.0 - Thu, 05 Nov 2015 22:29:26 GMT
--------------------------------------

- 


v0.2.1 - Fri, 23 Oct 2015 23:00:10 GMT
--------------------------------------

- [9f6d417](../../commit/9f6d417) [fixed] Removing ReactDOM from bundle


v0.2.0 - Thu, 22 Oct 2015 09:32:14 GMT
--------------------------------------

- [a478e8d](../../commit/a478e8d) [changed] support for React 0.14.0


v0.1.2 - Wed, 13 May 2015 20:57:15 GMT
--------------------------------------

- [db92743](../../commit/db92743) [fixed] don't include React in dist


v0.1.1 - Wed, 13 May 2015 20:11:02 GMT
--------------------------------------

- [de9f58c](../../commit/de9f58c) [fixed] empty dist files


v0.1.0 - Tue, 28 Apr 2015 22:56:46 GMT
--------------------------------------

- 




lib/components/__tests__/Tray-test.js
import React from 'react';
import ReactDOM from 'react-dom';
import TestUtils from 'react-addons-test-utils';
import { equal } from 'assert';
import ReactTray from '../../main';

let _currentDiv = null;

function renderTray(props, children, callback) {
  _currentDiv = document.createElement('div');
  document.body.appendChild(_currentDiv);
  return ReactDOM.render(<ReactTray {...props}>{children}</ReactTray>, _currentDiv, callback);
}

function unmountTray() {
  ReactDOM.unmountComponentAtNode(_currentDiv);
  document.body.removeChild(_currentDiv);
  _currentDiv = null;
}

/* eslint func-names:0 */
describe('react-tray', function() {
  afterEach(function() {
    unmountTray();
  });

  it('should not be visible when isOpen is false', function() {
    renderTray();
    equal(document.querySelectorAll('.ReactTray__Content').length, 0);
  });

  it('should be visible when isOpen is true', function() {
    renderTray({isOpen: true});
    equal(document.querySelectorAll('.ReactTray__Content').length, 1);
  });

  it('should receive focus when opened', function() {
    renderTray({isOpen: true});
    equal(document.querySelector('.ReactTray__Content'), document.activeElement);
  });

  it('should call onBlur when closed', function() {
    const blurred = false;
    renderTray({isOpen: true, onBlur: function() { blurred: true; }, closeTimeoutMS: 0});
    TestUtils.Simulate.click(document.querySelector('.ReactTray__Overlay'));
    setTimeout(function() {
      equal(blurred, true);
    }, 0);
  });

  it('should call onOpen when it opens', function() {
    let calledOpen = false;
    renderTray({isOpen: true, onOpen: function() { calledOpen = true;}});
    equal(calledOpen, true);
  });

  it('should close on overlay click', function() {
    renderTray({isOpen: true, onBlur: function() {}, closeTimeoutMS: 0});
    TestUtils.Simulate.click(document.querySelector('.ReactTray__Overlay'));
    setTimeout(function() {
      equal(document.querySelectorAll('.ReactTray__Content').length, 0);
    }, 0);
  });

  it('should close on ESC key', function() {
    renderTray({isOpen: true, onBlur: function() {}, closeTimeoutMS: 0});
    TestUtils.Simulate.keyDown(document.querySelector('.ReactTray__Content'), {key: 'Esc'});
    setTimeout(function() {
      equal(document.querySelectorAll('.ReactTray__Content').length, 0);
    }, 0);
  });

  it('should close on blur by default', function() {
    renderTray({isOpen: true, onBlur: function() {}, closeTimeoutMS: 0});
    TestUtils.Simulate.keyDown(document.querySelector('.ReactTray__Content'), {key: 'Tab'});
    setTimeout(function() {
      equal(document.querySelectorAll('.ReactTray__Content').length, 0);
    }, 0);
  });

  it('should not close on blur', function() {
    renderTray({isOpen: true, onBlur: function() {}, closeTimeoutMS: 0, closeOnBlur: false});
    TestUtils.Simulate.keyDown(document.querySelector('.ReactTray__Content'), {key: 'Tab'});
    setTimeout(function() {
      equal(document.querySelectorAll('.ReactTray__Content').length, 1);
    }, 0);
  });

  describe('maintainFocus prop', function() {
    this.timeout(0);
    beforeEach(function(done) {
      const props = {isOpen: true, onBlur: function() {}, closeTimeoutMS: 0, maintainFocus: true};
      const children = (
        <div>
          <a href="#" id="one">One</a>
          <a href="#" id="two">Two</a>
          <a href="#" id="three">Three</a>
        </div>
      );
      renderTray(props, children, () => done());
    });

    it('sends focus to the first item if tabbing away from the last element', function() {
      const firstItem = document.querySelector('#one');
      const lastItem = document.querySelector('#three');
      lastItem.focus();
      TestUtils.Simulate.keyDown(document.querySelector('.ReactTray__Content'), {keyCode: 9});
      equal(document.activeElement, firstItem);
    });

    it('sends focus to the last item if shift + tabbing from the first item', function() {
      const firstItem = document.querySelector('#one');
      const lastItem = document.querySelector('#three');
      firstItem.focus();
      TestUtils.Simulate.keyDown(document.querySelector('.ReactTray__Content'), {keyCode: 9, shiftKey: true});
      equal(document.activeElement, lastItem);
    });
  });

  describe('elementToFocus prop', function() {
    beforeEach(function(done) {
      const props = {isOpen: true, onBlur: function() {}, closeTimeoutMS: 0, maintainFocus: true, elementToFocus: '#two'};
      const children = (
        <div>
          <a href="#" id="one">One</a>
          <a href="#" id="two">Two</a>
          <a href="#" id="three">Three</a>
        </div>
      );
      renderTray(props, children, () => done());
    });

    it('sends focus to the DOM node found via the selector passed in the prop', function() {
      const secondItem = document.querySelector('#two');
      equal(document.activeElement, secondItem);
    });
  });

  describe('getAriaHideElement prop', function() {
    const getAriaHideElement = () => {
      return document.getElementById('main_application_div');
    };

    beforeEach(function() {
      const mainDiv = document.createElement('div');
      mainDiv.id = 'main_application_div';
      document.body.appendChild(mainDiv);
    });

    it('adds aria-hidden to the given element when open', function() {
      renderTray({isOpen: true, getAriaHideElement: getAriaHideElement});
      const el = document.getElementById('main_application_div');
      equal(el.getAttribute('aria-hidden'), 'true');
    });

    it('removes aria-hidden from the given element when closed', function() {
      renderTray({isOpen: true, onBlur: function() {}, closeTimeoutMS: 0, getAriaHideElement: getAriaHideElement});
      TestUtils.Simulate.keyDown(document.querySelector('.ReactTray__Content'), {key: 'Esc'});
      setTimeout(function() {
        const el = document.getElementById('main_application_div');
        equal(el.getAttribute('aria-hidden'), null);
      }, 0);
    });
  });
});


lib/components/Tray.js
import React from 'react';
import ReactDOM from 'react-dom';
import TrayPortal from './TrayPortal';
import { a11yFunction } from '../helpers/customPropTypes';
const renderSubtreeIntoContainer = ReactDOM.unstable_renderSubtreeIntoContainer;

export default React.createClass({
  displayName: 'Tray',

  propTypes: {
    isOpen: React.PropTypes.bool,
    onBlur: React.PropTypes.func,
    onOpen: React.PropTypes.func,
    closeTimeoutMS: React.PropTypes.number,
    closeOnBlur: React.PropTypes.bool,
    maintainFocus: React.PropTypes.bool,
    elementToFocus: React.PropTypes.string,
    getAriaHideElement: a11yFunction
  },

  getDefaultProps() {
    return {
      isOpen: false,
      closeTimeoutMS: 0,
      closeOnBlur: true,
      maintainFocus: true
    };
  },

  componentDidMount() {
    this.node = document.createElement('div');
    this.node.className = 'ReactTrayPortal';
    document.body.appendChild(this.node);
    this.renderPortal(this.props);
  },

  componentWillReceiveProps(props) {
    this.renderPortal(props);
  },

  componentWillUnmount() {
    ReactDOM.unmountComponentAtNode(this.node);
    document.body.removeChild(this.node);
  },

  renderPortal(props) {
    delete props.ref;

    renderSubtreeIntoContainer(this, <TrayPortal {...props}/>, this.node);
  },

  render() {
    return null;
  }
});


lib/components/TrayPortal.js
import React, { PropTypes } from 'react';
import cx from 'classnames';
import focusManager from '../helpers/focusManager';
import isLeavingNode from '../helpers/isLeavingNode';
import findTabbable from '../helpers/tabbable';

const styles = {
  overlay: {
    position: 'fixed',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0
  },
  content: {
    position: 'absolute',
    background: '#fff'
  }
};

const CLASS_NAMES = {
  overlay: {
    base: 'ReactTray__Overlay',
    afterOpen: 'ReactTray__Overlay--after-open',
    beforeClose: 'ReactTray__Overlay--before-close'
  },
  content: {
    base: 'ReactTray__Content',
    afterOpen: 'ReactTray__Content--after-open',
    beforeClose: 'ReactTray__Content--before-close'
  }
};

function isChild(parent, child) {
  if (parent === child) {
    return true;
  }

  let node = child;
  /* eslint no-cond-assign:0 */
  while (node = node.parentNode) {
    if (node === parent) {
      return true;
    }
  }
  return false;
}

export default React.createClass({
  displayName: 'TrayPortal',

  propTypes: {
    className: PropTypes.string,
    overlayClassName: PropTypes.string,
    isOpen: PropTypes.bool,
    onBlur: PropTypes.func,
    onOpen: PropTypes.func,
    closeOnBlur: PropTypes.bool,
    closeTimeoutMS: PropTypes.number,
    children: PropTypes.any,
    maintainFocus: PropTypes.bool,
    getElementToFocus: PropTypes.func,
    getAriaHideElement: PropTypes.func
  },

  getInitialState() {
    return {
      afterOpen: false,
      beforeClose: false
    };
  },

  componentDidMount() {
    if (this.props.isOpen) {
      this.setFocusAfterRender(true);
      this.open();
    }
  },

  componentWillReceiveProps(props) {
    if (props.isOpen) {
      this.setFocusAfterRender(true);
      this.open();
    } else if (this.props.isOpen && !props.isOpen) {
      this.close();
    }
  },

  componentDidUpdate() {
    if (this.focusAfterRender) {
      if (this.props.getElementToFocus) {
        this.props.getElementToFocus().focus();
      } else {
        this.focusContent();
      }
      this.setFocusAfterRender(false);
    }
  },

  setFocusAfterRender(focus) {
    this.focusAfterRender = focus;
  },

  focusContent() {
    this.refs.content.focus();
  },

  toggleAriaHidden(element) {
    if (!element.getAttribute('aria-hidden')) {
      element.setAttribute('aria-hidden', true);
    } else {
      element.removeAttribute('aria-hidden');
    }
  },

  handleOverlayClick(e) {
    if (!isChild(this.refs.content, e.target)) {
      this.props.onBlur();
    }
  },

  handleContentKeyDown(e) {
    // Treat ESC as blur/close
    if (e.keyCode === 27) {
      this.props.onBlur();
    }

    // Keep focus inside the tray if maintainFocus is true
    if (e.keyCode === 9 && this.props.maintainFocus && isLeavingNode(this.refs.content, e)) {
      e.preventDefault();
      const tabbable = findTabbable(this.refs.content);
      const target = tabbable[e.shiftKey ? tabbable.length - 1 : 0];
      target.focus();
      return;
    }

    // Treat tabbing away from content as blur/close if closeOnBlur
    if (e.keyCode === 9 && this.props.closeOnBlur && isLeavingNode(this.refs.content, e)) {
      e.preventDefault();
      this.props.onBlur();
    }
  },

  open() {
    focusManager.markForFocusLater();
    this.setState({isOpen: true}, () => {
      if (this.props.onOpen) {
        this.props.onOpen();
      }
      if (this.props.getAriaHideElement) {
        this.toggleAriaHidden(this.props.getAriaHideElement());
      }
      this.setState({afterOpen: true});
    });
  },

  close() {
    if (this.props.closeTimeoutMS > 0) {
      this.closeWithTimeout();
    } else {
      this.closeWithoutTimeout();
    }
    if (this.props.getAriaHideElement) {
      this.toggleAriaHidden(this.props.getAriaHideElement());
    }
  },

  closeWithTimeout() {
    this.setState({beforeClose: true}, () => {
      setTimeout(this.closeWithoutTimeout, this.props.closeTimeoutMS);
    });
  },

  closeWithoutTimeout() {
    this.setState({
      afterOpen: false,
      beforeClose: false
    }, this.afterClose);
  },

  afterClose() {
    focusManager.returnFocus();
  },

  shouldBeClosed() {
    return !this.props.isOpen && !this.state.beforeClose;
  },

  buildClassName(which) {
    let className = CLASS_NAMES[which].base;
    if (this.state.afterOpen) {
      className += ' ' + CLASS_NAMES[which].afterOpen;
    }
    if (this.state.beforeClose) {
      className += ' ' + CLASS_NAMES[which].beforeClose;
    }
    return className;
  },

  render() {
    return this.shouldBeClosed() ? <div/> : (
      <div
        ref="overlay"
        style={styles.overlay}
        className={cx(
          this.buildClassName('overlay'),
          this.props.overlayClassName
        )}
        onClick={this.handleOverlayClick}
      >
        <div
          ref="content"
          style={styles.content}
          className={cx(
            this.buildClassName('content'),
            this.props.className
          )}
          onKeyDown={this.handleContentKeyDown}
          tabIndex="-1"
        >
          {this.props.children}
        </div>
      </div>
    );
  }
});


examples/basic/app.js
import React from 'react';
import ReactDOM from 'react-dom';
import Tray from '../../lib/main';
import cx from 'classnames';

const App = React.createClass({
  getInitialState() {
    return {
      orientation: 'left',
      isTrayOpen: false
    };
  },

  handleNavClick(e) {
    const type = e.target.getAttribute('data-type');
    this.openTray(type);
  },

  handleNavKeyPress(e) {
    if (e.which === 13 || e.which === 32) {
      const type = e.target.getAttribute('data-type');
      this.openTray(type);
    }
  },

  handleOrientationChange(e) {
    this.setState({
      orientation: e.target.value
    });
  },

  renderTrayContent() {
    switch (this.state.type) {
    case 'foo':
      return (
        <div>
          <h2>Foo</h2>
          <div>Content for foo</div>
          <nav role="navigation">
            <div><a href="#">A</a></div>
            <div><a href="#">B</a></div>
            <div><a href="#">C</a></div>
          </nav>
        </div>
      );
    case 'bar':
      return (
        <div>
          <h2>Bar</h2>
          <div>Lorem Ipsum</div>
          <nav role="navigation">
            <div><a href="#">A</a></div>
            <div><a href="#">B</a></div>
            <div><a href="#">C</a></div>
          </nav>
        </div>
      );
    case 'baz':
      return (
        <div>
          <h2>Baz</h2>
          <div>Other stuff here</div>
          <nav role="navigation">
            <div><a href="#">A</a></div>
            <div><a href="#">B</a></div>
            <div><a href="#">C</a></div>
          </nav>
        </div>
      );
    default:
      return (
        <h1>You shouldn't see me</h1>
      );
    }
  },

  render() {
    return (
      <div>
        <ul role="menu" className={cx({
          'navigation': true,
          'navigation-left': this.state.orientation === 'left',
          'navigation-right': this.state.orientation === 'right'
        })}
        >
          <li role="menuitem"
              className={cx({ active: this.state.type === 'foo' })}
          >
            <a tabIndex={0}
                role="button"
                aria-haspopup={true}
                data-type="foo"
                onKeyPress={this.handleNavKeyPress}
                onClick={this.handleNavClick}>Foo</a>
          </li>
          <li role="menuitem"
              className={cx({ active: this.state.type === 'bar' })}
          >
            <a tabIndex={0}
                role="button"
                aria-haspopup={true}
                data-type="bar"
                onKeyPress={this.handleNavKeyPress}
                onClick={this.handleNavClick}>Bar</a>
          </li>
          <li role="menuitem"
              className={cx({ active: this.state.type === 'baz' })}
          >
            <a tabIndex={0}
                role="button"
                aria-haspopup={true}
                data-type="baz"
                onKeyPress={this.handleNavKeyPress}
                onClick={this.handleNavClick}>Baz</a>
          </li>
        </ul>
        <Tray isOpen={this.state.isTrayOpen}
              onBlur={this.closeTray}
              closeTimeoutMS={150}
              className={cx({
                'tray-left': this.state.orientation === 'left',
                'tray-right': this.state.orientation === 'right'
              })}
        >
          {this.renderTrayContent()}
        </Tray>
        <div className="content">
          <select onChange={this.handleOrientationChange}>
            <option value="left">Left</option>
            <option value="right">Right</option>
          </select>
        </div>
      </div>
    );
  },

  openTray(type) {
    this.setState({
      type: type,
      isTrayOpen: true
    });
  },

  closeTray() {
    this.setState({
      isTrayOpen: false
    }, () => {
      setTimeout(() => {
        this.setState({
          type: null
        });
      }, 150);
    });
  }
});

ReactDOM.render(<App/>, document.getElementById('example'));
